python/ambassador/config/resourceprocessor.py

```diff
@@ -0,0 +1,671 @@
+from typing import Any, ClassVar, ContextManager, Dict, FrozenSet, Iterator, List, Mapping, Optional, Type, TypeVar, Set
+
+import collections
+import collections.abc
+import contextlib
+import dataclasses
+import datetime
+import itertools
+import json
+import logging
+
+import durationpy
+
+from .config import Config
+from .acresource import ACResource
+
+from ..utils import dump_yaml
+
+
+@dataclasses.dataclass
+class Location:
+    """
+    Represents a location for parsing.
+    """
+
+    filename: Optional[str] = None
+    ocount: int = 1
+
+    def mark_annotation(self) -> None:
+        if self.filename is None:
+            return
+        elif self.filename.endswith(':annotation'):
+            return
+
+        self.filename += ':annotation'
+
+    def __str__(self) -> str:
+        return f"{self.filename or 'anonymous YAML'}.{self.ocount}"
+
+
+class LocationManager:
+    """
+    Manages locations contextually.
+    """
+
+    previous: List[Location]
+    current: Location
+
+    def __init__(self) -> None:
+        self.previous = []
+        self.current = Location()
+
+    def push(self, filename: Optional[str] = None, ocount: int = 1) -> ContextManager[Location]:
+        current = Location(filename, ocount)
+        self.previous.append(self.current)
+        self.current = current
+
+        # This trick lets you use the return value of this method in a `with`
+        # statement. At the conclusion of the statement block, the location will
+        # automatically be popped from the stack.
+        @contextlib.contextmanager
+        def popper():
+            yield current
+            self.pop()
+
+        return popper()
+
+    def push_reset(self) -> ContextManager[Location]:
+        """
+        Like push, but simply resets ocount keeping the current filename. Useful
+        for changing resource types.
+        """
+        return self.push(filename=self.current.filename)
+
+    def pop(self) -> Location:
+        current = self.current
+        self.current = self.previous.pop()
+        return current
+
+
+@dataclasses.dataclass
+class ResourceEmission:
+    """
+    Represents an Ambassador resource emitted after processing fetched data.
+    """
+
+    object: dict
+    rkey: Optional[str] = None
+
+    @classmethod
+    def from_data(cls, kind: str, name: str, namespace: str = 'default', generation: int = 1,
+                  version: str = 'v2', labels: Dict[str, Any] = None, spec: Dict[str, Any] = None) -> ResourceEmission:
+        rkey = f'{name}.{namespace}'
+
+        ir_obj = {}
+        if spec:
+            ir_obj.update(spec)
+
+        ir_obj['apiVersion'] = f'getambassador.io/{version}'
+        ir_obj['name'] = name
+        ir_obj['namespace'] = namespace
+        ir_obj['kind'] = kind
+        ir_obj['generation'] = generation
+        ir_obj['labels'] = labels or {}
+
+        return cls(ir_obj, rkey)
+
+
+@dataclasses.dataclass(frozen=True)
+class KubernetesGVK:
```

Comment:
I'm not sure it's been used elsewhere in Ambassador code but this is a pretty common acronym in controller-runtime. Happy to use an alternative if you have one.